bundle/jsky.app.ot/src/main/scala/jsky/app/ot/ags/BagsManager.scala

```diff
@@ -25,219 +24,396 @@ import jsky.app.ot.gemini.tpe.TpePWFSFeature
-    * *and* the specified program is still on the watch list.
+  /** RunningEdited. This state corresponds to a running AGS search for an
+    * observation that was subsequently edited.  Since it has been edited, the
+    * resuls we're expecting may no longer be valid when they arrive.
     */
-  private def dequeue(key: SPNodeKey, pid: SPProgramID): Boolean =
-    synchronized {
-      if (state.keys(key)) {
-        state -= key
-        state.programs(pid)
-      } else false
+  case class RunningEditedState(obs: ISPObservation, hash: AgsHashVal) extends BagsState {
+    // If we're edited again while running, just loop back.  Once the results of
+    // the previous edit that got us into RunningState in the first place
+    // finish, we'll switch to Pending and go again.
+    override val edit: StateTransition =
+      (this, ioUnit)
+
+    // Success, but now the observation has been edited so the AGS selection
+    // may not be valid.  Apply the results anyway in case the edit is not
+    // to a field that impacts AGS and go to pending to run again if necessary.
+    override def succeed(results: Option[AgsStrategy.Selection]): StateTransition = {
+      val action = for {
+        _ <- BagsManager.applyAction(obs, results)
+        _ <- BagsManager.wakeUpAction(obs, 0)
+      } yield ()
+      (PendingState(obs, Some(hash)), action)
     }
 
-  /**
-    * Atomically enqueue a task that will consider the specified observation for BAGS lookup, after
-    * a delay of at least `delay` milliseconds.
+    // Failed AGS but we've been edited in the meantime anyway.  Go back to
+    // PendingState so we can run again.  Here we pass in no AgsHash value to
+    // ensure that pending will count this as an AGS-worthy edit.
+    override def fail(why: String, delayMs: Long): StateTransition =
+      (PendingState(obs, None), BagsManager.wakeUpAction(obs, delayMs))
+  }
+
+  /** Failure. This state is used to mark an AGS lookup failure.  It is a
+    * transient state since a timer is expected to eventually go off and move
+    * us back to pending to retry the search.
     */
-  def enqueue(observation: ISPObservation, delay: Long): Unit = {
-    // Performs checks to rule out disabled guide groups, instruments without guiding strategies (e.g. GPI),
-    // observation classes that do not have guiding (e.g. daytime calibration).
-    def isEligibleForBags(ctx: ObsContext): Boolean = {
-      val enabledGroup = ctx.getTargets.getGuideEnvironment.guideEnv.auto match {
-        case AutomaticGroup.Initial | AutomaticGroup.Active(_,_) => true
-        case _                                                   => false
-      }
-      enabledGroup &&
-        AgsRegistrar.validStrategies(ctx).nonEmpty &&
-        ObsClassService.lookupObsClass(observation) != ObsClass.DAY_CAL
+  case class FailureState(obs: ISPObservation, why: String) extends BagsState {
```

Comment:
I should probably log it.  It was put there for the feedback originally of course but we're not giving feedback yet.